docs: report (#204)
Slug-Boi-ITU-Repositories/Dense-Dino-Nuggets: report/MSc_group_d.md

# DevOps Group D

| Name | Email |
| -------- | ------- |
| August Kofoed Brandt | <aubr@itu.dk> |
| Emilia Victoria Helsted | <ehel@itu.dk> |
| Niklas Zeeberg Hessner Christensen | <nizc@itu.dk> |
| Philip Guozhi Han Pedersen | <phgp@itu.dk> |
| Theis Per Holm | <thph@itu.dk> |

## 1. System's Perspective
### 1.2 Design and Architecture

![Allocation diagram of the Minitwit system](images/allocation-diagram.png)

> The allocation diagram shows how software elements are mapped to platform elements in the environment of the system. There are 2 droplets running on digital ocean. One droplet runs the minitwit app container. The minitwit application is run as 3 replicas using docker swarm, on the same droplet. All replicas don't need to be on this one droplet. If another droplet is part of the swarm, then a minitwit replica can be started on that droplet. We ran all replicas on the same droplet because we were limited to 3 droplets by GitHub Education, the second droplet was running our monitoring setup, which was using most of the available RAM on that droplet, and we were using the the 3rd droplet for testing. The swarm manager droplet also contains the minitwit database. The datafiles for the database are stored on an attached DigitalOcean volume allowing the files to persists if the droplet dies. The application containers running on the droplet uses a Loki logging driver acting as a middleware by sending logs to the Loki aggregator running on the monitoring/logging droplet, before writing the logs to standard output. Metrics from the app are pulled by the Prometheus container also running on the monitoring/logging droplet. Certbot is set up to get a TLS certificate from let's encrypt to enable HTTPS. The Nginx reverse proxy uses the certificate to enable HTTPS for the site and redirects HTTP requests to HTTPS.  No-IP Dynamic Update Client(DUC) is used to keep the domain name synchronized with the droplet IP. UFW is used to configure the firewall for the droplet.

Reviewer: ```suggestion
The allocation diagram shows how software elements are mapped to platform elements in the environment of the system. There are 2 droplets running on digital ocean. One droplet runs the minitwit app container. The minitwit application is run as 3 replicas using docker swarm, on the same droplet. All replicas don't need to be on this one droplet. If another droplet is part of the swarm, then a minitwit replica can be started on that droplet. We ran all replicas on the same droplet because we were limited to 3 droplets by GitHub Education, the second droplet was running our monitoring setup, with little RAM left on that droplet, and we were using the 3rd droplet for testing. The swarm manager droplet also contains the minitwit database. The datafiles for the database are stored on an attached DigitalOcean volume allowing the files to persists if the droplet dies. The application containers use a Loki logging driver acting as a middleware by sending logs to the Loki aggregator running on the monitoring/logging droplet, before writing the logs to standard output. Metrics from the app are pulled by the Prometheus container also running on the monitoring/logging droplet. Certbot is set up to get a TLS certificate from let's encrypt to enable HTTPS. The Nginx reverse proxy uses the certificate to enable HTTPS for the site and redirects HTTP requests to HTTPS.  No-IP Dynamic Update Client(DUC) is used to keep the domain name synchronized with the droplet IP. UFW is used to configure the firewall for the droplet.
```
Some more removals and refinements for improved readability

The third and last droplet runs a test environment and is not shown in the diagram above. It was used to test things like ansible and docker swarm before running it on our production environment.

The whole system can be booted up through OpenTofu and Ansible, which is able to create the digital ocean droplets and volume with OpenTofu, and configure and run the application through Ansible.

### 1.3 Dependencies

Our Minitwit uses the following dependencies:

#### Core dependencies

| Dependency | Description |
| ---------- | ----------- |
| Golang | Programming language. |
| PostgreSQL | Relational database. |
| GORM | ORM library for Go. |
| gorilla/mux | HTTP router and dispatcher. |
| godotenv | Module for loading .env files. |
| Nginx | Acts as a reverse proxy. |
| ufw | Firewall management. |
| Docker | Enables containerization of system components and orchestration of nodes using Docker swarm. |
| DigitalOcean | A cloud infrastructure provider. |
| OpenTofu | An infrastructure-as-code tool serving as an alternative to Terraform. |
| Certbot | A tool for automatically obtaining and renewing TLS certificates. |

> | No-IP | A Dynamic DNS provider that keeps the domain name in sync with the droplet's IP address using a DUC running on the server. |

Author: ```suggestion
| No-IP | Dynamic DNS provider. |
| DUC | Dynamic IP updater for No-IP. |
```

| Ansible | Automation tool for provisioning and configuring infrastructure through declarative playbooks. |
| Vagrant | **No longer a dependency.**  A tool for building and provisioning development environments using virtual machines. |

#### Testing and Quality tooling

| Dependency | Description |
| ---------- | ----------- |
| Dagger | Workflow orchestration. |
| Playwright | A web-facing end-to-end testing library. |
| misspell | Spellchecker. |
| golangci-lint | A universal linter. |
| CodeQL | Static code analysis, focused on security and vulnerability detection. |
| SonarQube | Static code analysis, focused on code quality and maintainability. |
| Codacy | Aggregates results of static code analysis and presents in PR's. |

#### Monitoring and Logging

| Dependency | Description |
| ---------- | ----------- |
| Grafana | Metric and log visualization. |
| Loki | Log aggregation system and logging driver. |
| Prometheus | Timeseries database for metrics |
| Prometheus Client Library | Library for collecting and exposing monitoring data |

### 1.4 Current State of our Minitwit

#### Security

> The group did a security assessment of the system and concluded the following.

Reviewer 2: ```suggestion
We did a security assessment of the system and concluded the following:
```

- The most critical risks involve the backend API and sensitive credentials
- Infrastructure misconfiguration presents significant risk
- Proper security controls significantly reduce overall risk exposure

> The background for these findings was outlined in security assessment which can be found through the link in the appendix.

Reviewer 2: ```suggestion
The background for these findings was outlined in security assessment, which can be found through the link in the appendix.
```

> Additionally for security we used dockerscout to find vulnerabilities in our dockerimage. The initial run resulted in 40 total vulnerabilities 4 of them being high severity.

Reviewer 2: ```suggestion
Additionally for security we used Docker Scout to find vulnerabilities in our Docker image. The initial run resulted in a total of 40 vulnerabilities with 4 of them being of high severity.
```

![Dockerscout vulnerabilities before](images/dockerscout_CVE_before.png)

> After fixing these we are down to  vulnerabilities

Reviewer 2: There seems to be a number missing???

Author: ```suggestion
After fixing these we are down to one vulnerability 
```

![Dockerscout vulnerabilities before](images/dockerscout_CVE_after.png)

> Lastly we added dependabot very late to the project and as such still have open security issues based on this which can be seen on the Github repository.

Reviewer 2: ```suggestion
We added Dependabot late into the project and still have open security issues, which can be seen on the Github repository.
```

#### Code Quality

> The project used a few static analysis tools for checking code periodically and on Pull Requests. Codacy was one of these tools and is currently passing on Main. When Codacy found issues on PR's they were fixed before merging into Main. Another static analysis tool we used was Sonarqube Cloud, this one is currently not passing on Main. Sonarqube was extremely badly configured for our project and was therefore very useless for most PR's. Most of the security problems it outlined were due to us not using TLs at the time and most of the code duplication it pointed out was false flags (playwright tests had 80+% code duplication). This tool could have been very useful it we had taken more time to configure it properly but as it stands the tool was mostly a hindrance.

Reviewer 2: I low key wanna rewrite this to make it more concise

#### Testing

We refactored the existing test suite from the original version of Minitwit and added end-to-end tests to test the UI parts of the web app. We did not have tests for the API endpoints, as an attempt to mitigate this we have ran the simulator periodically but this should have been a part of the test suite in some way.

#### IaC vs Current State

The monitoring droplet currently does not mount the services that are running on it as docker volumes. This would need to be added to the docker-compose file for the data to be accessible for the different containers. Previously we had set this up by mounting a file path (when we were still using vagrant), this was not mirrored in the migration to OpenTofu. Other than this small discrepancy the IaC should match the current system.

## 2. Process' perspective
### 2.1 CI/CD Pipelines

#### Validation pipeline on pull requests

> Our PR pipeline is triggered on pull requests and pushes to main to ensure the quality of the code merged into main. When a developer creates a pull request a series of automated quality and security checks are initiated. We run SonarQube, CodeQL, and Codeacy for static code analysis. We also run our test workflow, as seen in the diagram below, which runs our Go tests, linting, and spellchecker misspell, all orchestrated through Dagger. SonarQube and Codacy both post a report on the pull request for a quick overview. We do manual peer reviews where the other developers can suggest changes. We require that all the checks pass and at least two members of our team review and approve the changes in the pull request. When both of these conditions are met the pull request can be merged into main.

Author: ```suggestion
Our PR pipeline is triggered on pull requests and pushes to main to ensure the quality of the code merged into main. When a developer creates a pull request a series of automated quality and security checks are initiated. We also run our test workflow, as seen in the diagram below, which runs our Go tests, linting, and spellchecker misspell, all orchestrated through Dagger. SonarQube and Codacy both post a report on the pull request for a quick overview. We do manual peer reviews where the other developers can suggest changes. We require that all the checks pass and at least two members of our team review and approve the changes in the pull request. 
```

![test_CI_pipeline.png](images/test_CI_pipeline_test.png)

#### Release Pipeline

> Below is a flowchart showing the Release CI pipeline. It is triggered when a version tag is pushed. Dagger orchestrates and runs our Go tests, linting, spellcheck, and Playwright-based end-to-end tests. If all tests succeed we build a packaged release artifact and publish a Docker image, making the new version of Minitwit ready for deployment.

Author: ```suggestion
Below is a flowchart showing the Release CI pipeline. It is triggered when a version tag is pushed. Dagger orchestrates and runs our Go tests, linting, spellcheck, and Playwright-based end-to-end tests. If all tests succeed we build a packaged release artifact and publish a Docker image, and provision the new image to our DO droplet.
```

![release_CI_pipeline.png](images/test_CI_pipeline_release.png)

### 2.2 Monitoring of Minitwit

#### Dashboard structure

This is how our monitoring dashboard looks
![minitwit_dashboard.png](./images/minitwit_dashboard.png)

Our monitoring dashboard tracks the following:

- If the server is running or is down

> - The 99'th and 99.9'th percentile of response time to see if we have requests that are taking longer than they should

Reviewer 2: ```suggestion
- The 99'th and 99.9'th percentile of response time
```

- The average time requests take to know how the system is responding in general
- The amount of requests per second to see the load on the system
- The logs from minitwit to see what is happening on the system

#### Alerting

> We set up an alert that used Grafana's built-in Discord web hook, such that a Discord bot would send a message in our Discord server in case the minitwit system went down or was unreachable by Prometheus. We tested this artificially but never in practice since the system did not go down after we added the alert.

Reviewer 2: ```suggestion
We set up an alert that used Grafana's built-in Discord web hook, such that a Discord bot would send a message in our Discord server in case the minitwit system went down or was unreachable by Prometheus.
```

#### Monitoring issues

> One large issue we ran into was that our Prometheus setup was not collecting accurate information from the minitwit system, after we started using swarm to have multiple replicas. This was an unfortunate side effect of Prometheus being a pull system. Because, when Prometheus tried to pull data it would only get the monitoring data from one of the replicas. To solve this we needed to move over to a push system, where each of the replicas would need a system to push their monitoring data to Prometheus such that it would be able to aggregate data from all the replicas, or show data for each replica. However, we did not have the time to implement this as there were other things that took higher priority, so this never reached the top of the priority list.

Reviewer 2: ```suggestion
After we started using swarm to have multiple replicas our Prometheus setup stopped collecting accurate information from the Minitwit system. This was an unfortunate side effect of Prometheus being a pull system, since when Prometheus tried to pull data it would only get the monitoring data from one of the replicas. To solve this we need to move over to a push system, where each of the replicas would need a system to push their monitoring data to Prometheus so it can be aggregated or the data for each replica could be shown separately. We did however not have the time to implement this.
```

Author: ```suggestion
After we started using swarm to have multiple replicas our Prometheus setup stopped collecting accurate information from the Minitwit system. This was an unfortunate side effect of Prometheus being a pull system, since when Prometheus tried to pull data it would only get the monitoring data from one of the replicas. To solve this we need to move over to a push system, where each of the replicas would need a system to push their monitoring data to Prometheus so it can be aggregated and shown separately in grafana. We did however not have the time to implement this.
```

### 2.3 Logging

> For logging we had a logging middleware that we used for web requests, which logged what type of request is was, what endpoint it was too, and how long the request took to execute. We also used logging in parts of the code where an error could occur to give information about what went wrong. Gorm also has it's own logging which show information about the query that was executed, when there are issues or warnings. Like the Query taking too long or not returning any rows. One big issue with our logging is that we used simple `log.Printf` lines for all our logging. This means we didn't have the ability have different levels of logging or filtering the logs based on the type. In practice this didn't lead to any issues since we didn't run into any time where the server ran into errors after having started successfully. Any time we ran into errors they where while the servers was starting and the logs, therefore were easy to look through.

Reviewer 2: ```suggestion
We have a logging middleware that we use for web requests, which logs type of request, endpoint, and execution time of request. We also use logging in error-prone parts of the code for diagnostics. Gorm also has its own logging, which shows information about queries with issues or warnings. An issue with our logging is that we use simple `log.Printf` lines for all our logging. This means the we do not have the ability have different levels of logging or filtering of the logs based on type. In practice this did not lead to any issues, since we did not run into any issues after having started Minitwit successfully. 
```

Author: ```suggestion
We have logging middleware that we use for web requests, which logs type of request, endpoint, and execution time of requests. We also use logging in error-prone parts of the code for diagnostics. Gorm also has its own logging, which shows information about queries with issues or warnings. An issue with our logging is that we use simple `log.Printf` lines for all our logging. This means that we don't have the ability have different levels of logging or filtering of the logs based on type. In practice this didn't lead to any issues, since we didn't run into any issues after having started Minitwit successfully. 
```

remove a and change the to that, did not -> didn't, do not -> don't

When looking at our backlog we see that we have many stale issues and a few duplicates. As the backlog grew the Github project board became less reliable as a planning tool and showed that we would need a better planning structure if we were to continue on with the project.
Ultimately we identified a lot of small non-technical issues early on but did not do much to mitigate them. This all shows that awareness is not enough and we need to actually dedicate time to address them.

A significant change with this project is the fact that we were responsible for operations after deployment. The project required a shift to a continuous delivery mindset where we had to focus more on quality and maintainability. Our extended feedback loop ensured that we were constantly aware of these qualities through pull requests and monitoring, rather than focusing solely on feature delivery.

 The whole process came with a higher need for coordination than previous projects due to shifting demands and complexity.

## 4. Use of Generative AI

> Our project includes AI-assisted code. We have used the following generative models while working on our Minitwit:

Reviewer 3: ```suggestion
We made use of the following generative models for our work on Minitwit:
```

- ChatGPT
- DeepSeek
- Claude

> Throughout the course we have used generative AI to explain code and help us understand different topics and technologies, so we could more easily work with them. When we refactored the tests from the original Minitwit, ChatGPT was used to help debug and identify why one of the tests was failing, so we could fix the issue. We have also used generative AI for code generation and to aid us in refactoring parts of the project. While we have made some effort to co-author the AI models when it was used, we have not tracked our AI usage systematically, so we are not able to identify all specific instances where generated code has been included.

Reviewer 3: ```suggestion
We have used generative AI to explain code and help us understand different topics and technologies, so we could more easily work with them. We also used AI for code generation and to aid us in refactoring parts of the project. While we have made some effort to co-author the AI models when it was used, we have not tracked our AI usage systematically, so we are not able to identify all specific instances where generated code has been included.
```

> While generative AI helped us with certain tasks and accelerated the process, it also introduced technical debt when its code was included without being properly reviewed. An example was when we refactored from Vagrant to Ansible. Since the vagrantfile consisted of over 500 lines of shell script code, we used ChatGPT to translate it into a playbook. It translated the code too literally and did not take advantage of many Ansible specific features. Due to this we had to spend a substantial amount of time on fixing and extending the generated code.

Reviewer 3: ```suggestion
While generative AI helped us with certain tasks and accelerated the process, it also introduced technical debt when its code was included without being properly reviewed. An example was when we refactored from Vagrant to Ansible. Since the vagrantfile consisted of over 500 lines of shell script code, we used ChatGPT to translate it into a playbook. It translated the code too literally and did not take advantage of many Ansible specific features. Due to this we had to spend a substantial amount of time on fixing the generated code.
```

| Loki | Log aggregation system and logging driver. |
| Prometheus | Timeseries database for metrics |
| Prometheus Client Library | Library for collecting and exposing monitoring data |

### 1.4 Current State of our Minitwit

#### Security

We did a security assessment of the system and concluded the following:

- The most critical risks involve the backend API and sensitive credentials
- Infrastructure misconfiguration presents significant risk
- Proper security controls significantly reduce overall risk exposure

The background for these findings was outlined in security assessment, which can be found through the link in the appendix.

Additionally for security we used Docker Scout to find vulnerabilities in our Docker image. The initial run resulted in a total of 40 vulnerabilities with 4 of them being of high severity.

![Dockerscout vulnerabilities before](images/dockerscout_CVE_before.png)

After fixing these we are down to one vulnerability 

![Dockerscout vulnerabilities before](images/dockerscout_CVE_after.png)

We added Dependabot late into the project and still have open security issues, which can be seen on the Github repository.

#### Code Quality

> The project used a few static analysis tools for checking code periodically and on Pull Requests. Codacy was one of these tools and is currently passing on Main. When Codacy found issues on PR's they were fixed before merging into Main. Another static analysis tool we used was Sonarqube Cloud, this one is currently not passing on Main. Sonarqube was extremely badly configured for our project and was therefore very useless for most PR's. Most of the security problems it outlined were due to us not using TLs at the time and most of the code duplication it pointed out was false flags (playwright tests had 80+% code duplication). This tool could have been very useful it we had taken more time to configure it properly but as it stands the tool was mostly a hindrance.

Reviewer 2: ```suggestion
To ensure code quality, we use the static code analysis tools Codacy, SonarQube Cloud, and CodeQL on pull requests. This mitigates misspelled words, code duplication, security issues, etc.  SonarQube is currently not passing on our main branch. It was poorly configured for our project and did not add a lot of value to our process. All other checks pass.
```

| No-IP | Dynamic DNS provider. |
| DUC | Dynamic IP updater for No-IP. |
| Ansible | Automation tool for provisioning and configuring infrastructure through declarative playbooks. |
| Vagrant | **No longer a dependency.**  A tool for building and provisioning development environments using virtual machines. |

#### Testing and Quality tooling

| Dependency | Description |
| ---------- | ----------- |
| Dagger | Workflow orchestration. |
| Playwright | A web-facing end-to-end testing library. |
| misspell | Spellchecker. |
| golangci-lint | A universal linter. |
| CodeQL | Static code analysis, focused on security and vulnerability detection. |
| SonarQube | Static code analysis, focused on code quality and maintainability. |
| Codacy | Aggregates results of static code analysis and presents in PR's. |

#### Monitoring and Logging

| Dependency | Description |
| ---------- | ----------- |
| Grafana | Metric and log visualization. |
| Loki | Log aggregation system and logging driver. |
| Prometheus | Timeseries database for metrics |
| Prometheus Client Library | Library for collecting and exposing monitoring data |

### 1.4 Current State of our Minitwit

#### Security

> We did a security assessment of the system and concluded the following:

Reviewer: ```suggestion
 A security assessment of the system revealed the following:
```

- The most critical risks involve the backend API and sensitive credentials
- Infrastructure misconfiguration presents significant risk
- Proper security controls significantly reduce overall risk exposure

> The background for these findings was outlined in security assessment, which can be found through the link in the appendix.

Reviewer: ```suggestion
The background for these findings was outlined in the security assessment (link in appendix).
```

> Additionally for security we used Docker Scout to find vulnerabilities in our Docker image. The initial run resulted in a total of 40 vulnerabilities with 4 of them being of high severity.

Reviewer: ```suggestion
We used Docker Scout to find vulnerabilities in our Docker image. The initial run resulted in a total of 40 vulnerabilities with 4 of them being of high severity.
```

2. Convert the testing droplet into a droplet that is part of the production setup. This would allow us to set all 3 droplets as managers, which would let us maintain a manager if one droplet goes down.
3. Setup another Postgres service on a different droplet to act as a backup using Postgres streaming replication. Then the system can switch to this database if the main database goes down.

The reason we have not implemented these solutions is a combination of lack of time, and the need to have a testing droplet for testing our migration to Ansible and OpenTofu. We were limited to 3 droplets by the fact that we were using GitHub education.

## 3. Reflection Perspective
<!---
Describe the biggest issues, how you solved them, and which are major lessons learned with regards to:

- evolution and refactoring
- operation
- maintenance
of your ITU-MiniTwit systems. Link back to respective commit messages, issues, tickets, etc. to illustrate these.

Also reflect and describe what was the "DevOps" style of your work. For example, what did you do differently to previous development projects and how did it work?
-->

> As the course progressed challenges of our project slowly became less about implementing new features and more about handling the growing complexity of the project.

Author: ```suggestion
As the course progressed challenges of our project slowly became less about implementing new features and more about handling the growing complexity of the system.
```

> We started suffering from knowledge silos early, which we mitigated by introducing meetings each Friday, where we talked about what we had worked on throughout the week and how it works. It is evident that documentation was more of an afterthought and not sufficient for sharing knowledge across the team. Communication is thus clearly an important part of infrastructure for ensuring maintainability.

Author: ```suggestion
We started suffering from knowledge silos early, which we mitigated by introducing meetings each Friday, where we talked about what we had worked on throughout the week. It is evident that documentation was more of an afterthought and not sufficient for sharing knowledge across the team. Communication is thus clearly an important part of infrastructure for ensuring maintainability.
```

> When looking at our backlog we see that we have many stale issues and a few duplicates. As the backlog grew the kanban board became less reliable as a planning tool and showed that we would need a better planning structure if we were to continue on with the project.

Author: > As the backlog grew the kanban board

Unsure if we should refer to it as a kanban board here. We have not used it as a kanban board so I would prefer we find a different word for it

Author: ```suggestion
When looking at our backlog we see that we have many stale issues and a few duplicates. As the backlog grew the Github project board became less reliable as a planning tool and showed that we would need a better planning structure if we were to continue on with the project.
```

credit @Flakiator for suggestion

> Ultimately we identified a lot of these small non-technical issues early on but did not do much to mitigate them. This all shows that awareness is not enough and we need to actually dedicate time to address them.

Author: ```suggestion
Ultimately we identified a lot of small non-technical issues early on but did not do much to mitigate them. This all shows that awareness is not enough and we need to actually dedicate time to address them.
```

what are these? Suggest removing

> A significant change with this project is the fact that we were responsible for operations after deployment. The project required a shift to a continuous delivery mindset where we had to focus more on quality and maintainability than previously. Our extended feedback loop ensured that we were constantly aware of these qualities through pull requests and monitoring, rather than focusing solely on feature delivery.

Author: ```suggestion
A significant change with this project is the fact that we were responsible for operations after deployment. The project required a shift to a continuous delivery mindset where we had to focus more on quality and maintainability. Our extended feedback loop ensured that we were constantly aware of these qualities through pull requests and monitoring, rather than focusing solely on feature delivery.
```

what is exnded feedback loop? Is this mentioned earlier if not I suggest finding another word for this or rewriting to explain

Reviewer 2: Extended feedback loop consists of our pr pipeline, code checks, peer reviews, monitoring etc.

>  The whole process came with a higher need for coordination than previous projects due to shifting demands and complexity.

Author: This part is hard to find a union for us all. I wouldn't say this is the case for me at least

Reviewer: I guess we could call it a "github project board"?

Reviewer 2: I think it depends on what year one took BDSA. I can only relate it to ISE. I don't remember my BDSA covering any of this besides having to refactor a lot (but it was also ages ago)

<!---
Lil' thoughts:

"Awareness is not resolution"
 
Technical debt in the form of Process debt.

Generative AI adds/reduces technical debt -> Not tried that before. Belongs to GenAI part

From the DevOps handbook:
- We have a high lead time

-->

## 4. Use of Generative AI
<!---
ITU's rules on the use of generative AI apply for this report too. They are described https://itustudent.itu.dk/Study%20Administration/Generative%20AI#Guidelines and in detail https://itustudent.itu.dk/-/media/ITU-Student/Study-Administration/GAI/Generative-AI-guidelines-for-students-Spring-2026-pdf.pdf. Please follow them. For your report that means that you have to state which generative AI tools have been used for which task(s) in your projects. Additionally, describe how generative AI tools have been used and briefly reflect and discuss how they supported or hindered your work process.

From the guidelines:
• State which generative AI technology has been used.
• Describe how generative AI technology has been used.
-->
Our project includes AI-assisted code. We have used the following generative models while working on our Minitwit:

- ChatGPT
- DeepSeek
- Claude

> Throughout the course we have used generative AI to explain code and help us understand different topics and technologies, so we could more easily work with them. When we refactored the tests from the original Minitwit, ChatGPT was used to help debug and identify why one of the tests was failing, so we could fix the issue. We have also used generative AI for code generation and to aid us in refactoring parts of the project. We have not tracked our AI usage systematically, so we are not able to identify all specific instances where generated code has been included.

Author: ```suggestion
Throughout the course we have used generative AI to explain code and help us understand different topics and technologies, so we could more easily work with them. When we refactored the tests from the original Minitwit, ChatGPT was used to help debug and identify why one of the tests was failing, so we could fix the issue. We have also used generative AI for code generation and to aid us in refactoring parts of the project. We have not tracked our AI usage systematically, so we are not able to identify all specific instances where generated code has been included, however we have still made an effort to co-author AI in cases where it was used.
```

Is this not just wrong? For instances of AI use we should have co-authored them so that would count no?

Reviewer 2: I have made another suggestion with a similar change.

> While generative AI helped us with certain tasks and accelerated the process, it also introduced technical debt when its code was included without being properly reviewed. An example was when we refactored from Vagrant to Ansible. Since the vagrantfile consisted of over 500 lines of code, we used ChatGPT to translate it into a playbook. It translated the code too literally and did not take advantage of the way Ansible works. Due to this we had to spend a substantial amount of time on fixing the generated code.

Author: ```suggestion
While generative AI helped us with certain tasks and accelerated the process, it also introduced technical debt when its code was included without being properly reviewed. An example was when we refactored from Vagrant to Ansible. Since the vagrantfile consisted of over 500 lines of shell script code, we used ChatGPT to translate it into a playbook. It translated the code too literally and did not take advantage of many Ansible specific features. Due to this we had to spend a substantial amount of time on fixing and extending the generated code.
```

> Introducing AI generated code into our codebase also meant that we had code that we did not necessarily understand, which is bad for maintainability. In general we should have been more critical of our use of generative AI as it takes away from our learning when it does our work for us.

Author: > Introducing AI generated code into our codebase also meant that we had code that we did not necessarily understand, which is bad for maintainability.

Do we have code in the code base we don't understand? My use case of AI I was reviewing the output and changing things. The compose file might have had small things but I think that's about it. Unsure if this should be changed unless somebody else can come up with a concrete case for this

<!---
Just some flowy thoughts:
I belive we accidently set up CodePilot reviews for some pull requests

> As a group, we did not establish a formal policy or set of guidelines for the use of generative AI during the project, and usage was therefore informal and unregulated. Generative AI was used by multiple team members for tasks such as code generation, debugging, refactoring, and explanation of technical concepts. Since usage was not systematically documented at commit or task level, it is not possible to precisely attribute which parts of the codebase were AI-assisted. However, all contributions, including AI-assisted code, were reviewed and integrated by the team as part of the normal development workflow.

Author: I like this part honestly could just move it out of the comment into the report. The last part I still don't know that i fully agree with unless any group member has not documented this. Again for my part i have co-authored when I used AI code so unsure if it was just me or if this part could just get deleted

<!---
Lil' thoughts:

"Awareness is not resolution"
 
Technical debt in the form of Process debt.

Generative AI adds/reduces technical debt -> Not tried that before. Belongs to GenAI part

From the DevOps handbook:
- We have a high lead time

-->

## 4. Use of Generative AI
<!---
ITU's rules on the use of generative AI apply for this report too. They are described https://itustudent.itu.dk/Study%20Administration/Generative%20AI#Guidelines and in detail https://itustudent.itu.dk/-/media/ITU-Student/Study-Administration/GAI/Generative-AI-guidelines-for-students-Spring-2026-pdf.pdf. Please follow them. For your report that means that you have to state which generative AI tools have been used for which task(s) in your projects. Additionally, describe how generative AI tools have been used and briefly reflect and discuss how they supported or hindered your work process.

From the guidelines:
• State which generative AI technology has been used.
• Describe how generative AI technology has been used.
-->
Our project includes AI-assisted code. We have used the following generative models while working on our Minitwit:

- ChatGPT
- DeepSeek
- Claude

> Throughout the course we have used generative AI to explain code and help us understand different topics and technologies, so we could more easily work with them. When we refactored the tests from the original Minitwit, ChatGPT was used to help debug and identify why one of the tests was failing, so we could fix the issue. We have also used generative AI for code generation and to aid us in refactoring parts of the project. We have not tracked our AI usage systematically, so we are not able to identify all specific instances where generated code has been included.

Reviewer 2: ```suggestion
Throughout the course we have used generative AI to explain code and help us understand different topics and technologies, so we could more easily work with them. When we refactored the tests from the original Minitwit, ChatGPT was used to help debug and identify why one of the tests was failing, so we could fix the issue. We have also used generative AI for code generation and to aid us in refactoring parts of the project. While we have made some effort to co-author the AI models when it was used, we have not tracked our AI usage systematically, so we are not able to identify all specific instances where generated code has been included.
```

# DevOps Group D

| Name | Email |
| -------- | ------- |
| August Kofoed Brandt | <aubr@itu.dk> |
| Emilia Victoria Helsted | <ehel@itu.dk> |
| Niklas Zeeberg Hessner Christensen | <nizc@itu.dk> |
| Philip Guozhi Han Pedersen | <phgp@itu.dk> |
| Theis Per Holm | <thph@itu.dk> |

## 1. System's Perspective
<!---
Here we need:
A description and illustration of the:

- Design and architecture of your ITU-MiniTwit systems.

- All dependencies of your ITU-MiniTwit systems on all levels of abstraction and development stages. That is, list and briefly describe all technologies and tools you applied and depend on.

- Describe the current state of your systems, for example using results of static analysis and quality assessments.
-->
### 1.2 Design and Architecture

![Allocation diagram of the Minitwit system](images/allocation-diagram.png)

> The allocation diagram shows how software elements are mapped to platform elements in the environment of the system. There are 2 droplets running on digital ocean. One droplet runs the containerized version of the minitwit application. The minitwit application is run as 3 replicas using docker swarm, all running on the same droplet. The replicas don't need to be on this one droplet. If another droplet is part of the swarm, then a minitwit replica can be started on that droplet. The reason we are running all the replicas on one droplet is that we where limited to 3 droplets by GitHub Education, the second droplet was running our monitoring setup, which was using most of the available RAM on that droplet, and we were using the the 3rd droplet to have a testing droplet. The swarm manager droplet also contains the minitwit database. The datafiles for the database are stored on an attached DigitalOcean volume that allows the files to persists if the droplet dies. The application containers running on the droplet uses a Loki logging driver which acts as a middleware by sending logs to the Loki aggregator running on the monitoring/logging droplet, before writing the logs to standard output. Metrics from the app are pulled by the Prometheus container also running on the monitoring/logging droplet. Certbot is set up to get a TLS certificate from let's encrypt to enable HTTPS. The Nginx reverse proxy uses the certificate to enable HTTPS for the site and redirects HTTP requests to HTTPS.  No-IP Dynamic Update Client(DUC) is run on a docker container to keep the domain name synchronized with the droplet IP. UFW is used to configure the firewall for the droplet.

Reviewer: ```suggestion
The allocation diagram shows how software elements are mapped to platform elements in the environment of the system. There are 2 droplets running on digital ocean. One droplet runs the minitwit app container. The minitwit application is run as 3 replicas using docker swarm, on the same droplet. All replicas don't need to be on this one droplet. If another droplet is part of the swarm, then a minitwit replica can be started on that droplet. We ran all replicas on the same droplet because we were limited to 3 droplets by GitHub Education, the second droplet was running our monitoring setup, which was using most of the available RAM on that droplet, and we were using the the 3rd droplet for testing. The swarm manager droplet also contains the minitwit database. The datafiles for the database are stored on an attached DigitalOcean volume allowing the files to persists if the droplet dies. The application containers running on the droplet uses a Loki logging driver acting as a middleware by sending logs to the Loki aggregator running on the monitoring/logging droplet, before writing the logs to standard output. Metrics from the app are pulled by the Prometheus container also running on the monitoring/logging droplet. Certbot is set up to get a TLS certificate from let's encrypt to enable HTTPS. The Nginx reverse proxy uses the certificate to enable HTTPS for the site and redirects HTTP requests to HTTPS.  No-IP Dynamic Update Client(DUC) is used to keep the domain name synchronized with the droplet IP. UFW is used to configure the firewall for the droplet.
```
Tried to shorten some of the sentences went from 285 to 267.

![Dockerscout vulnerabilities before](images/dockerscout_CVE_before.png)

After fixing these we are down to  vulnerabilities

![Dockerscout vulnerabilities before](images/dockerscout_CVE_after.png)

Lastly we added dependabot very late to the project and as such still have open security issues based on this which can be seen on the Github repository.

#### Code Quality

The project used a few static analysis tools for checking code periodically and on Pull Requests. Codacy was one of these tools and is currently passing on Main. When Codacy found issues on PR's they were fixed before merging into Main. Another static analysis tool we used was Sonarqube Cloud, this one is currently not passing on Main. Sonarqube was extremely badly configured for our project and was therefore very useless for most PR's. Most of the security problems it outlined were due to us not using TLs at the time and most of the code duplication it pointed out was false flags (playwright tests had 80+% code duplication). This tool could have been very useful it we had taken more time to configure it properly but as it stands the tool was mostly a hindrance.

#### Testing

> We translated the existing test suite from the Python version of the program and added end-to-end tests to test the UI parts of the webapp. We did not have tests for the API endpoints, as an attempt to mitigate this we have ran the simulator periodically but this should have been a part of the test suite in some way.

Reviewer 2: ```suggestion
We refactored the existing test suite from the original version of Minitwit and added end-to-end tests to test the UI parts of the web app. We did not have tests for the API endpoints, as an attempt to mitigate this we have ran the simulator periodically but this should have been a part of the test suite in some way.
```

#### IaC vs Current State

> The Monitoring droplet currently does not mount the services that are running on it as docker volumes this would need to be added to the docker-compose file for the data to be accessible for the different containers. In the past we had set this up mounting a file path (when we were still using vagrant), this was not mirrored in the migration to OpenTofu. Other than this small discrepancy the IaC should match the current system.

Reviewer 2: ```suggestion
The monitoring droplet currently does not mount the services that are running on it as docker volumes. They would need to be added to the docker-compose file for the data to be accessible for the different containers. Previously we had set this up by mounting a file path (when we were still using vagrant), this was not mirrored in the migration to OpenTofu. Other than this small discrepancy the IaC should match the current system.
```

Author: ```suggestion
The monitoring droplet currently does not mount the services that are running on it as docker volumes. This would need to be added to the docker-compose file for the data to be accessible for the different containers. Previously we had set this up by mounting a file path (when we were still using vagrant), this was not mirrored in the migration to OpenTofu. Other than this small discrepancy the IaC should match the current system.
```

-->

#### Release Pipeline

Below is a flowchart showing the Release CI pipeline. It is triggered when a version tag is pushed. Dagger orchestrates and runs our Go tests, linting, spellcheck, and Playwright-based end-to-end tests. If all tests succeed we build a packaged release artifact and publish a Docker image, making the new version of Minitwit ready for deployment.

<!--
We need to make the release diagram and fix the placeholder here!!!!
-->

![release_CI_pipeline.png](images/test_CI_pipeline_release.png)

### 2.2 Monitoring of Minitwit

#### Dashboard structure

This is how our monitoring dashboard looks
![minitwit_dashboard.png](./images/minitwit_dashboard.png)

> Data is gathered from the minitwit application by pulling with Prometheus. Logs are aquired by using the Loki logging driver on our minitwit containers, which push to an aggregator container on the monitoring machine. Grafana can then pull data from Prometheus and Loki to display in the dashboard. The dashboard shows:

Reviewer: ```suggestion
Our monitoring dashboard tracks the following:
```
Maybe delete some of the first sentences and just focus on what we monitor? Otherwise this sounds too much like repetition of what we already wrote in the design and architecture chapter.

Throughout the course we have used generative AI to explain code and help us understand different topics and technologies, so we could more easily work with them. When we refactored the tests from the original Minitwit, ChatGPT was used to help debug and identify why one of the tests was failing, so we could fix the issue. We have also used generative AI for code generation and to aid us in refactoring parts of the project. While we have made some effort to co-author the AI models when it was used, we have not tracked our AI usage systematically, so we are not able to identify all specific instances where generated code has been included.

While generative AI helped us with certain tasks and accelerated the process, it also introduced technical debt when its code was included without being properly reviewed. An example was when we refactored from Vagrant to Ansible. Since the vagrantfile consisted of over 500 lines of shell script code, we used ChatGPT to translate it into a playbook. It translated the code too literally and did not take advantage of many Ansible specific features. Due to this we had to spend a substantial amount of time on fixing and extending the generated code.

> Introducing AI generated code into our codebase also meant that we had code that we did not necessarily understand, which is bad for maintainability. In general we should have been more critical of our use of generative AI as it takes away from our learning when it does our work for us.

Reviewer 2: ```suggestion
In general we should have been more critical of our use of generative AI as it sometimes became a bit of a crutch and took away from our learning when it did our work for us.
```
